app/services/control_rate_service.rb

```diff
@@ -0,0 +1,93 @@
+class ControlRateService
+  PERCENTAGE_PRECISION = 1
+
+  # Can be initialized with _either_ a date range or a single date to calculate
+  # control rates. Note that for the date range the returned values will be for each month going back
+  # to the beginning of registrations for the region.
+  def initialize(region, range: nil, date: nil)
+    raise ArgumentError, "Cannot provide both a range and date" if range && date
+    raise ArgumentError, "Must provide either a range or a single date" if range.nil? && date.nil?
+    @region = region
+    @facilities = if @region.respond_to?(:facilities)
+      @region.facilities
+    else
+      [region]
+    end
+    @range = range
+    @date = date
+    @end_of_date_range = date || range.end
+    logger.info "#{self.class} created for range: #{range} region: #{region.id} #{region.name}"
+  end
+
+  delegate :logger, to: Rails
+  attr_reader :date
+  attr_reader :facilities
+  attr_reader :range
+  attr_reader :region
+
+  def call
+    data = {
+      controlled_patients: {},
+      controlled_patients_rate: {},
+      registrations: {}
+    }
+
+    data[:cumulative_registrations] = registrations(@end_of_date_range)
+    registration_counts.each do |(date, count)|
+      formatted_period = date.to_s(:month_year)
+      data[:controlled_patients][formatted_period] = controlled_patients(date).count
+      data[:controlled_patients_rate][formatted_period] = percentage(controlled_patients(date).count, count)
+      data[:registrations][formatted_period] = count
+    end
+    data
+  end
+
+  def registrations(time)
+    registration_counts[time.beginning_of_month.to_date]
+  end
+
+  # Calculate all registration counts for entire range, or for the single date provided
+  def registration_counts
+    @registration_counts ||= if range
+      region.patients.with_hypertension
+        .group_by_period(:month, :recorded_at, range: range)
+        .count
+        .each_with_object(Hash.new(0)) { |(date, count), hsh|
+          hsh[:running_total] += count
+          hsh[date] = hsh[:running_total]
+        }.delete_if { |date, count| count == 0 }.except(:running_total)
+    else
+      count = region.patients.with_hypertension.where("recorded_at <= ?", date).count
+      {
+        date => count
+      }
+    end
+  end
+
+  def controlled_patients(time)
+    end_range = time.end_of_month
+    mid_range = time.advance(months: -1).end_of_month
+    beg_range = time.advance(months: -2).end_of_month
+    # We need to avoid the default scope to avoid ambiguous column errors, hence the `with_discarded`
+    # Note that the deleted_at scoping piece is applied when the SQL view is created, so we don't need to worry about it here
+    sub_query = LatestBloodPressuresPerPatientPerMonth
+      .with_discarded
```

Comment:
Is `.unscoped` a more semantic way to achieve the same thing?

Reply:
My thought was: If it's specifically the discarded scope that's throwing off this query, then `with_discarded` is good. If it's _any_ default scope throwing this thing off, then I'd prefer `unscoped`, as what we're really looking to do is reset.

Not a huge deal, I'll let you take a call! 👍

Reply:
Ahh, I see, yeah I think the only thing throwing this off was the `with_discarded` piece. 👍